rflx/model.py

```diff
@@ -372,10 +405,116 @@ def __prove_contradictions(self) -> None:
+                    raise ModelError(f'negative length for field "{f.name}" on path {path_message}'
+                                     f' ({result}: {message})')
+
+                first = l.first if l.first != UNDEFINED else self.__link_first(l)
+                start = If([(And(self.__type_constraints(And(path_expressions, first)),
+                                 path_expressions),
+                             GreaterEqual(first, First('Message')))],
+                           TRUE)
+                result = start.forall()
+                if result != ProofResult.sat:
+                    path_message = ' -> '.join([l.target.name for l in p])
+                    message = str(start.simplified()).replace('\n\t', '')
+                    raise ModelError(f'start of field "{f.name}" on path {path_message} before'
+                                     f' message start ({result}: {message})')
+
+    def __prove_coverage(self) -> None:
+        """
+        Prove that the fields of a message cover all message bits, i.e. there are no holes in the
+        message definition.
+
+        Idea: Let f be the bits covered by the message. By definition
+            (1) f >= Message'First and f <= Message'Last
+        holds. For every field add a conjunction of the form
+            (2) Not(f >= Field'First and f <= Field'Last),
+        effectively pruning the range that this field covers from the bit range of the message. For
+        the overall expression, prove that it is false for all f, i.e. no bits are left.
+        """
+        for path in [p[:-1] for p in self.__paths[FINAL] if p]:
+            # Calculate (1)
+            message_range = And(GreaterEqual(Variable('f'), First('Message')),
+                                LessEqual(Variable('f'), Last('Message')))
+            # Calculate (2) for all fields
+            fields = And(*[Not(And(GreaterEqual(Variable('f'), self.__link_first(l)),
+                                   LessEqual(Variable('f'), self.__link_last(l)))) for l in path])
+            # Define that the end of the last field of a path is the end of the message
+            last_field = Equal(self.__link_last(path[-1]), Last('Message'))
+            # Constraints for links and types
+            path_expressions = self.__with_constraints(And(*[self.__link_expression(l)
+                                                             for l in path]))
+
+            # Coverage expression must be False, i.e. no bits left
+            coverage = Not(And(*[fields, last_field, path_expressions, message_range]))
+            result = coverage.forall()
+            if result != ProofResult.sat:
+                path_message = ' -> '.join([l.target.name for l in path[:-1]])
+                message = str(coverage).replace('\n\t', '')
+                raise ModelError(f'path {path_message} does not cover whole message'
+                                 f' ({result}: {message})')
```

Comment:
I implemented a unit test, but I have trouble getting it checked by mypy. The problem is, that the coverage issue will be sorted out by other tests of the Message class. To avoid that, I derived my own test class from Message, and overrod `__init__` and called only part of the tests. The problem is that I need to call internal class members of the base class. This can be done by using `self._Message__method()` in the derived class, but mypy complains that the derived class does not have such a method.

@treiher Do you have other ideas how to solve that?

Reply:
I think the more sophisticated approach is to mock the method which catches the issue first. I guess in this case it should be sufficient to replace `__verify_conditions` by an empty method:

```
with mock.patch('rflx.model.Message._Message__verify_conditions', lambda x: None):
```

(If you want to have a look at an example, `mock` is used in `test_cli.py`.)

Reply:
Witchcraft! That's exactly what I was looking for, thanks! Fixed in dfefeda.